caffe2/CMakeLists.txt

```diff
@@ -787,1208 +817,1383 @@ else()
+  elseif(USE_ROCM)
+    target_compile_definitions(torch_hip PRIVATE TORCH_HIP_BUILD_MAIN_LIB)
+  endif()
 
-caffe2_interface_library(torch_cpu torch_cpu_library)
+  if(USE_XPU)
+    target_compile_definitions(torch_xpu PRIVATE TORCH_XPU_BUILD_MAIN_LIB)
+  endif()
 
-if(USE_CUDA)
-  caffe2_interface_library(torch_cuda torch_cuda_library)
-elseif(USE_ROCM)
-  caffe2_interface_library(torch_hip torch_hip_library)
-elseif(USE_XPU)
-  caffe2_interface_library(torch_xpu torch_xpu_library)
-endif()
+  set(EXPERIMENTAL_SINGLE_THREAD_POOL
+      "0"
+      CACHE
+        STRING
+        "Experimental option to use a single thread pool for inter- and intra-op parallelism"
+  )
+  if("${EXPERIMENTAL_SINGLE_THREAD_POOL}")
+    target_compile_definitions(torch_cpu
+                               PUBLIC "-DAT_EXPERIMENTAL_SINGLE_THREAD_POOL=1")
+  endif()
 
-caffe2_interface_library(torch torch_library)
+  if(MSVC AND BUILD_SHARED_LIBS)
+    # ONNX is linked statically and needs to be exported from this library to be
+    # used externally. Make sure that references match the export.
+    target_compile_options(torch_cpu PRIVATE "-DONNX_BUILD_MAIN_LIB")
+  endif()
 
-install(TARGETS torch_cpu torch_cpu_library EXPORT Caffe2Targets DESTINATION "${TORCH_INSTALL_LIB_DIR}")
+  caffe2_interface_library(torch_cpu torch_cpu_library)
 
-if(USE_CUDA)
-  install(TARGETS torch_cuda torch_cuda_library EXPORT Caffe2Targets DESTINATION "${TORCH_INSTALL_LIB_DIR}")
-elseif(USE_ROCM)
-  install(TARGETS torch_hip torch_hip_library EXPORT Caffe2Targets DESTINATION "${TORCH_INSTALL_LIB_DIR}")
-elseif(USE_XPU)
-  install(TARGETS torch_xpu torch_xpu_library EXPORT Caffe2Targets DESTINATION "${TORCH_INSTALL_LIB_DIR}")
-endif()
+  if(USE_CUDA)
+    caffe2_interface_library(torch_cuda torch_cuda_library)
+  elseif(USE_ROCM)
+    caffe2_interface_library(torch_hip torch_hip_library)
+  elseif(USE_XPU)
+    caffe2_interface_library(torch_xpu torch_xpu_library)
+  endif()
 
-install(TARGETS torch torch_library EXPORT Caffe2Targets DESTINATION "${TORCH_INSTALL_LIB_DIR}")
+  caffe2_interface_library(torch torch_library)
 
-target_link_libraries(torch PUBLIC torch_cpu_library)
+  install(
+    TARGETS torch_cpu torch_cpu_library
+    EXPORT Caffe2Targets
+    DESTINATION "${TORCH_INSTALL_LIB_DIR}")
 
-if(USE_CUDA)
-  target_link_libraries(torch PUBLIC torch_cuda_library)
-elseif(USE_ROCM)
-  target_link_libraries(torch PUBLIC torch_hip_library)
-endif()
+  if(USE_CUDA)
+    install(
+      TARGETS torch_cuda torch_cuda_library
+      EXPORT Caffe2Targets
+      DESTINATION "${TORCH_INSTALL_LIB_DIR}")
+  elseif(USE_ROCM)
+    install(
+      TARGETS torch_hip torch_hip_library
+      EXPORT Caffe2Targets
+      DESTINATION "${TORCH_INSTALL_LIB_DIR}")
+  elseif(USE_XPU)
+    install(
+      TARGETS torch_xpu torch_xpu_library
+      EXPORT Caffe2Targets
+      DESTINATION "${TORCH_INSTALL_LIB_DIR}")
+  endif()
 
-if(USE_XPU)
```

Comment:
leaving off manual review here. outcome seems clear -- it's a little clunky and there are a few spots we could improve but it seems generally fine